member_exprt now checks type of compound operand

kroening opened the pull request:
This commit adds a precondition to the constructors of `member_exprt` that
enforces that the compound operand has one of the compound types.

A legacy constructor is offered to ease the transition.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig reviewed (commented):
It seems the need for `unchecked` exclusively stems from tests? Can we just fix them instead?

kroening committed:
member_exprt now checks type of compound operand

This commit adds a precondition to the constructors of member_exprt that
enforces that the compound operand has one of the compound types.

tautschnig commented on unit/solvers/smt2_incremental/object_tracking.cpp:
```diff
@@ -55,7 +59,7 @@ TEST_CASE("Tracking object base expressions", "[core][smt2_incremental]")
     notequal_exprt{
       address_of_exprt{
-        member_exprt{foo, "baz", unsignedbv_typet{8}}, pointer_type},
+        member_exprt::unchecked(foo, "baz", unsignedbv_typet{8}), pointer_type},
```
This will still need fixing?

peterschrammel commented on unit/solvers/smt2_incremental/object_tracking.cpp:
```diff
@@ -12,41 +12,12 @@
 #include <string>
 #include <utility>
 
-TEST_CASE("find_object_base_expression", "[core][smt2_incremental]")
```
@thomasspriggs is going to re-add a fixed version of this test after this PR has been merged.
